pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,850 @@
+	}
+	if err := bc.buildDefaults.ApplyDefaults(podSpec); err != nil {
+		return nil, fmt.Errorf("failed to apply build defaults for build %s/%s: %v", build.Namespace, build.Name, err)
+	}
+	if err := bc.buildOverrides.ApplyOverrides(podSpec); err != nil {
+		return nil, fmt.Errorf("failed to apply build overrides for build %s/%s: %v", build.Namespace, build.Name, err)
+	}
+	return podSpec, nil
+}
+
+// resolveOutputDockerImageReference returns a reference to a Docker image
+// computed from the buid.Spec.Output.To reference.
+func (bc *BuildController) resolveOutputDockerImageReference(build *buildapi.Build) (string, error) {
+	outputTo := build.Spec.Output.To
+	if outputTo == nil || outputTo.Name == "" {
+		return "", nil
+	}
+	var ref string
+	switch outputTo.Kind {
+	case "DockerImage":
+		ref = outputTo.Name
+	case "ImageStream", "ImageStreamTag":
+		// TODO(smarterclayton): security, ensure that the reference image stream is actually visible
+		namespace := outputTo.Namespace
+		if len(namespace) == 0 {
+			namespace = build.Namespace
+		}
+
+		var tag string
+		streamName := outputTo.Name
+		if outputTo.Kind == "ImageStreamTag" {
+			var ok bool
+			streamName, tag, ok = imageapi.SplitImageStreamTag(streamName)
+			if !ok {
+				return "", fmt.Errorf("the referenced image stream tag is invalid: %s", outputTo.Name)
+			}
+			tag = ":" + tag
+		}
+		stream, err := bc.imageStreamStore.ImageStreams(namespace).Get(streamName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return "", fmt.Errorf("the referenced output image stream %s/%s does not exist", namespace, streamName)
+			}
+			return "", fmt.Errorf("the referenced output image stream %s/%s could not be found by build %s/%s: %v", namespace, streamName, build.Namespace, build.Name, err)
+		}
+		if len(stream.Status.DockerImageRepository) == 0 {
+			e := fmt.Errorf("the image stream %s/%s cannot be used as the output for build %s/%s because the integrated Docker registry is not configured and no external registry was defined", namespace, outputTo.Name, build.Namespace, build.Name)
+			bc.recorder.Eventf(build, kapi.EventTypeWarning, "invalidOutput", "Error starting build: %v", e)
+			return "", e
+		}
+		ref = fmt.Sprintf("%s%s", stream.Status.DockerImageRepository, tag)
+	}
+	return ref, nil
+}
+
+// createBuildPod creates a new pod to run a build
+func (bc *BuildController) createBuildPod(build *buildapi.Build) (*buildUpdate, error) {
+
+	update := &buildUpdate{}
+
+	// Set the output Docker image reference.
+	ref, err := bc.resolveOutputDockerImageReference(build)
+	if err != nil {
+		update.setReason(buildapi.StatusReasonInvalidOutputReference)
+		update.setMessage(buildapi.StatusMessageInvalidOutputRef)
+		return update, err
+	}
+
+	// Create the build pod spec
+	buildPod, err := bc.createPodSpec(build, ref)
+	if err != nil {
+		update.setReason(buildapi.StatusReasonCannotCreateBuildPodSpec)
```

Comment:
The createPodSpec is not telling the rest api to create a pod on the master. It's simply constructing a pod object in memory that corresponds to the build that we passed it. Like I said, the main reason it would fail is if the build has a strategy that we don't understand.